r-lib-error-1.0.md

```diff
@@ -0,0 +1,60 @@
+
+# Specification of r-lib errors, version 1.0
+
+Note: this is an experimental draft.
+
+Note: version 1.0 corresponds to the `rlang_error` class, as
+implemented in rlang 0.4.0.
+
+## Classes
+
+An r-lib error object always has at least classes `rlib_error`, `error`
+and `condition`. It may have other classes.
+
+## Members
+
+An `rlib_error` object has the following members:
+
+* `message`: `character(1)`, the error message.
+* `trace`: optional, an `rlib_trace` object, see below. 
+* `parent`: optional, pointer to a parent error, of class `error`. It
+  may also be `NULL`, which is the same as missing.
+* `call`: optional, the captured call, of class `call`.
+* `version`: optional, `r_lib` error version, defaults to 1.0, i.e. this
+  specification.
+
+## Traces
+
+An `rlib_trace` object has the following members:
+
+* `calls`: A list of calls, each of class `call`. This is the stack when
+  the error was thrown. `calls[[1]]` is the top of the stack.
+* `parents`: Integer vector of parent frame pointers, as in `sys.parents()`.
+* `envs`: List or character vector of environment names. They are typically
+  package names for frames that correspond to functions in packages.
+* `indices`: Frame numbers, typically just 1:n.
+* `version`: options, `r_lib` error version, defaults to 1.0, i.e. this
+  specification.
+
+`calls`, `parents`, `envs` and `indices` must have the same length.
+
+## Parent errors
+
+Parent errors may be arbitrary error objects. A parent error may also
+be another `rlib_error`, that includes another trace, etc.
+
+If the error has a parent error (i.e. `parent` is not `NULL`), then
+the trace of the error may only include part of the stack: from the top to
```

Comment:
I think we should remove this paragraph. The child error does not really include part of the stack, it's just that the parent error was captured with an exiting handler so the stack has unwound by that point.

Reply:
We can remove it from 1.0, but eventually I want to be able include the full stack, to avoid having to print `err$parent$parent$...$trace` to see the full stack.

Reply:
I'm not concerned about this since we can just have `cnd_root()` to pick up the last trace. Whatever we do, this will be the most reliable way to get the full stack, since some child error will be created in an exiting handler.

Reply:
An exiting handler does not have the stack any more. From an exiting handler you can only have the stack if you create it when you call `abort()`.

Reply:
Also, if I remove this, is it obvious that the ultimate error only shows part of the stack? It is not really obvious to me.

Reply:
That's what I meant, a child error created from an exiting handler won't have the full stack. I think we should leave it up to the error creator whether they use an exiting or a calling handler. In that case, the reliable way of accessing the full stack is `cnd_root()$trace`.

We're not really adding a new function because the spec implies how to get the root with a loop by subsetting `$parent` until reaching the root.

Reply:
> I think we should leave it up to the error creator whether they use an exiting or a calling handler. 

Well, considering that an exiting handler loses the stack, I would advise to use a non-exiting handler, whenever you are re-throwing the error. We don't have to specify this advice in this spec, but I think we need to say something about what kind of the frames can go into `trace` and `$parrent$trace`. Or can we just say they can be whatever frames? Doesn't that make it too hard to write a print method for example?

> In that case, the reliable way of accessing the full stack is cnd_root()$trace.

I don't think that's true. If I `stop()` and then `tryCatch()` then there is no way to recover the stack, right? Or may there is a trick I don't know?

Reply:
> I don't think that's true. If I stop() and then tryCatch() then there is no way to recover the stack, right? Or may there is a trick I don't know?

`abort()` saves the trace before throwing, which I agree is not optimal.